Add mutual TLS docs

kilfoyle opened the pull request:
Thanks for the [nice draft](https://docs.google.com/document/d/1oU-_I4BsUBRpG42aEEp8Fq2Q7PgyMEHrS7w6eWN8MgY/edit) @nimarezainia! I made only very, very small changes.

@AndersonQ and @michel-laterman  Let me know if you're not the right folks to review this. I'm not sure who worked on the feature.

Please see [Docs Preview](https://ingest-docs_bk_1132.docs-preview.app.elstc.co/guide/en/fleet/master/mutual-tls.html)

There are also some mTLS docs changes in https://github.com/elastic/ingest-docs/pull/1099

Closes: #1131

nimarezainia commented:
@michel-laterman @AndersonQ would you mind reviewing these to ensure accuracy with what we have in code. TIA

AndersonQ commented on docs/en/ingest-management/security/mutual-tls.asciidoc:
```diff
@@ -0,0 +1,200 @@
+[discrete]
+=== {fleet} settings
+
+This is the same as what's described for <<mutual-tls-on-premise,on premise deployments>>. The main difference is that you need to use certificates that are accepted by the secure proxy, as that is where the TLS connection will be terminated.
```
[Suggestion]
Personally I'd rather focus more on the mTLS than the TLS termination itself. Also because you can have TLS termination without mTLS. Something in the line of:

```suggestion
This is the same as what's described for <<mutual-tls-on-premise,on premise deployments>>. The main difference is that you need to use certificates that are accepted by the secure proxy, as the mTLS is setup between the agent and the secure proxy.
```

kilfoyle replied:
Changed to:

> This is the same as what's described for <<mutual-tls-on-premise,on premise deployments>>. The main difference is that you need to use certificates that are accepted by the secure proxy, as the mTLS is set up between the agent and the secure proxy.

AndersonQ commented on docs/en/ingest-management/security/mutual-tls.asciidoc:
```diff
@@ -0,0 +1,200 @@
+
+[cols="1,1"]
+|===
+|`-certificate-authorities`
```
[Blocker]
the correct key is:
```suggestion
|`ssl.certificate-authorities`
```

kilfoyle replied:
I've changed all instances of `-certificate-authorities` to `ssl.certificate-authorities`.

AndersonQ commented on docs/en/ingest-management/images/mutual-tls-onprem-advanced-yaml.png:
I'd suggest to also add a `ssl.certificate-authorities` entry in the image

kilfoyle replied:
I updated the screencap advanced YAML settings to:

```
ssl.certificate-authorities: "/etc/pki/client/ca.crt"
ssl.certificate: "/etc/pki/client/cert.pem"
ssl.key: "/etc/pki/client/cert.key"
```

nimarezainia replied:
thanks. We will be changing these soon to have dedicated UI elements to enter the CA/certs etc

AndersonQ reviewed (changes requested):
Folks, here are my 2 (or 22) cents.

There are a few disparities between what is written and how things actually happen or are. Those I marked with `[Blocker]`, they need to be addressed or `[Blocker-ish]` I'd rather see them addressed, but if you understand my point and still believe the current version is good, I won't make a stand.

The ones marked with `[Suggestion]` are exactly that, suggestions, feel free to ignore, no need to explain to me why.

Also I've put some suggested changes, those are just suggestion, feel free to use whatever wording you find more appropriated. I just used it as a tool to make my point clear rather than asking it to b taken as I put.

kilfoyle commented on docs/en/ingest-management/security/mutual-tls.asciidoc:
```diff
@@ -0,0 +1,200 @@
+[[mutual-tls]]
+= {agent} deployment models with mutual TLS
+
+Mutual Transport Layer Security (TLS) provides a higher level of security and trust compared to one-way TLS, where only the server presents a certificate. It ensures that not only the server is who it claims to be, but the client is also authenticated. This is particularly valuable in scenarios where both parties need to establish trust and validate each other's identities, such as in secure API  ...
```
Changed to `mTLS`.

kilfoyle commented on docs/en/ingest-management/security/mutual-tls.asciidoc:
```diff
@@ -0,0 +1,200 @@
+
+* **Client Authentication**: The client presents its digital certificate to the server during the TLS handshake. This certificate is issued by a trusted Certificate Authority (CA) and contains the client's public key.
+* **Server Authentication**: The server also presents its digital certificate to the client, proving its identity and sharing its public key. The server's certificate is also issued by a trusted CA.
+* **Certificate Verification**: Both the client and server verify each other's certificates by checking the digital signatures against the CA's public key. This process confirms the authenticity of the certificates and, by extension, the identities of the client and server.
```
I think it's probably worth noting that the CAs need not be the same, so I've add that as a note.

> Both the client and server verify each other's certificates by checking the digital signatures against the CAs' public key (note that the client and server need not use the same CA).

kilfoyle commented on docs/en/ingest-management/security/mutual-tls.asciidoc:
```diff
@@ -0,0 +1,200 @@
+
+* **Client Authentication**: The client presents its digital certificate to the server during the TLS handshake. This certificate is issued by a trusted Certificate Authority (CA) and contains the client's public key.
+* **Server Authentication**: The server also presents its digital certificate to the client, proving its identity and sharing its public key. The server's certificate is also issued by a trusted CA.
+* **Certificate Verification**: Both the client and server verify each other's certificates by checking the digital signatures against the CA's public key. This process confirms the authenticity of the certificates and, by extension, the identities of the client and server.
```
I've removed that sentence.

kilfoyle commented on docs/en/ingest-management/security/mutual-tls.asciidoc:
```diff
@@ -0,0 +1,200 @@
+
+In order to bootstrap, {agent} initially must establish a secure connection to the {fleet-server}, which can reside on-premises or in {ecloud}. This connectivity verification process ensures the agent's authenticity. Once verified, the agent receives the policy configuration. This policy download equips the agent with the knowledge of the other components it needs to engage with. For instance, it ...
+
+If mutual TLS (mTLS) is a requirement, {agent} must first establish an mTLS connection with {fleet-server}, with both client and server exchanging certificates and validating one another. Once the policy configuration is in place, it possesses the necessary details to establish an mTLS connection with the specific output it's configured to use. In the case of {fleet}-managed {agents}, certificate ...
```
Sounds good! I've removed a couple of paragraphs and added the precedence:

![Screenshot 2024-06-19 at 3 35 10 PM](https://github.com/elastic/ingest-docs/assets/41695641/e5f1b173-f4e9-47d6-b2c0-84d6240841bd)

kilfoyle commented on docs/en/ingest-management/security/mutual-tls.asciidoc:
```diff
@@ -0,0 +1,200 @@
+
+|`-fleet-server-es-cert`
+|{fleet-server} certificate to present to {es}
+∫
```
Funny - I wouldn't know how to type that character if I wanted to. I can only do it by accident. :-)

kilfoyle commented on docs/en/ingest-management/security/mutual-tls.asciidoc:
```diff
@@ -0,0 +1,200 @@
+|`-fleet-server-es-cert-key`
+|{fleet-server} certificate key to present to {es}
+
+|`-certificate-authorities`
```
I've changed this to:

> -certificate-authorities
> List of CA certificates that are trusted when {agent} connects to {fleet-server} and when {fleet-server} validates the {agent} identity.

kilfoyle commented on docs/en/ingest-management/security/mutual-tls.asciidoc:
```diff
@@ -0,0 +1,200 @@
+|{fleet-server} certificate to present to {agents} during authentication
+
+|`-fleet-server-cert-key`
+|{fleet-server} certificate key to present to {agents}  
```
Changed to:

> |-fleet-server-cert-key
> {fleet-server's} private certificate key used to decrypt the certificate

I also made this change where the setting appears further down  in the "on-prem hosted ES" section.

kilfoyle commented on docs/en/ingest-management/security/mutual-tls.asciidoc:
```diff
@@ -0,0 +1,200 @@
+You may either apply the full certificate, in which case all the agents get the same certificate OR alternatively point to a local directory on the agent where the certificate resides, if the certificates are to be unique per agent.
+
+|`ssl.key`
+|This certificate key will be passed down to all the agents that have this output configured in their policy. This key is used by the agent when establishing mTLS to the output. 
```
Changed to:

> |`ssl.key`
> |This certificate key will be passed down to all the agents that have this output configured in their policy. The certificate key is used to decrypt the SSL certificate.

> Similar to the certificate, the certificate key may be stored in a local directory.

kilfoyle commented on docs/en/ingest-management/security/mutual-tls.asciidoc:
```diff
@@ -0,0 +1,200 @@
+
+[cols="1,1"]
+|===
+|`-certificate-authorities`
+|List of CA certificates that are trusted when {fleet-server} connects to {agent}
+
+|`ssl.certificate`
+|This certificate will be passed down to all the agents that have this output configured in their policy. This certificate is used by the agent when establishing mTLS to the output. 
+
+You may either apply the full certificate, in which case all the agents get the same certificate OR alternatively point to a local directory on the agent where the certificate resides, if the certificates are to be unique per agent.
+
+|`ssl.key`
+|This certificate key will be passed down to all the agents that have this output configured in their policy. This key is used by the agent when establishing mTLS to the output. 
```
I've put this in as a note, below the settings table:

<img width="692" alt="Screenshot 2024-06-19 at 4 14 44 PM" src="https://github.com/elastic/ingest-docs/assets/41695641/7ef63948-7378-45cf-8d49-215773fda308">

kilfoyle commented on docs/en/ingest-management/security/mutual-tls.asciidoc:
```diff
@@ -0,0 +1,200 @@
+[[mutual-tls-cloud]]
+== {fleet-server} on {ecloud}
+
+In this deployment model, all traffic ingress into {ecloud} has its TLS connection terminated at an Elastic-managed Cloud Proxy. Cloud Proxy is not provisioned on a per tenant basis so at this point a client specific certificate can NOT be used in this proxy.
```
Added exactly as suggested.

kilfoyle commented on docs/en/ingest-management/security/mutual-tls.asciidoc:
```diff
@@ -0,0 +1,200 @@
+
+[discrete]
+[[mutual-tls-cloud-proxy]]
+== {fleet-server} on {ecloud} using a proxy
```
I didn't change anything, as I think we use the same titles in other spots about the various deployment models, but this is really good to know!

kilfoyle commented on docs/en/ingest-management/security/mutual-tls.asciidoc:
```diff
@@ -0,0 +1,200 @@
+[cols="1,1"]
+|===
+|`-certificate-authorities`
+|List of CA certificates that are trusted when {fleet-server} connects to {agent} 
```
Changed to:

> ssl.certificate-authorities
> List of CA certificates that are trusted when {agent} connects to {fleet-server} or to the proxy between {agent} and {fleet-server}.

kilfoyle commented on docs/en/ingest-management/security/mutual-tls.asciidoc:
```diff
@@ -0,0 +1,200 @@
+[[mutual-tls-cloud-proxy]]
+== {fleet-server} on {ecloud} using a proxy
+
+In this scenario, unlike the one above, you have access to the proxy and therefore can configure it to do a TLS termination.
```
Fixed exactly as suggested.

kilfoyle commented on docs/en/ingest-management/security/mutual-tls.asciidoc:
```diff
@@ -0,0 +1,200 @@
+|`-certificate-authorities`
+|List of CA certificates that are trusted when {fleet-server} connects to {agent} 
+
+|`-elastic-agent-cert`
+|{agent} certificate to present to {fleet-server} during authentication
+
+|`-elastic-agent-cert-key`
+|{agent} certificate key to present to {fleet-server}
```
Changed to:

> -elastic-agent-cert
> {agent} certificate to present  during authentication to {fleet-server} or to the proxy between {agent} and {fleet-server}

> -elastic-agent-cert-key
> {agent}'s private certificate key used to decrypt the certificate

kilfoyle commented on docs/en/ingest-management/security/mutual-tls.asciidoc:
```diff
@@ -0,0 +1,200 @@
+[[mutual-tls-on-premise-hosted-es]]
+== {fleet-server} on-premise and Hosted Elasticsearch Service
+
+In some scenarios you may want to deploy {fleet-server} on your own premises. In this case, you're able to provide your own certificates and certificate authority to facilitate mTLS between {fleet-server} and {agent}.
```
Good point! I changed this to use `enable`:

> In some scenarios you may want to deploy {fleet-server} on your own premises. In this case, you're able to provide your own certificates and certificate authority to enable mTLS between {fleet-server} and {agent}.

kilfoyle commented on docs/en/ingest-management/security/mutual-tls.asciidoc:
```diff
@@ -0,0 +1,200 @@
+
+In some scenarios you may want to deploy {fleet-server} on your own premises. In this case, you're able to provide your own certificates and certificate authority to facilitate mTLS between {fleet-server} and {agent}.
+
+However, as with the <<mutual-tls-cloud,{fleet-server} on {ecloud}>> use case, the data plane connections are terminated at Elastic's Cloud Proxy, which is not a multi-tenanted service and therefore can't provide per-user certificates.  
```
I changed this to:

> However, as with the <<mutual-tls-cloud,{fleet-server} on {ecloud}>> use case, the data plane connections are has its TLS connection terminated at the {ecloud} boundary. {ecloud} is not a multi-tenanted service and therefore can't provide per-user certificates.

kilfoyle commented on docs/en/ingest-management/security/mutual-tls.asciidoc:
```diff
@@ -0,0 +1,200 @@
+[cols="1,1"]
+|===
+|`-certificate-authorities`
+|List of CA certificates that are trusted for when {fleet-server} connects to {agent} 
```
Changed to:

> ssl.certificate-authorities
> List of CA certificates that are trusted for when {agent} connects to {fleet-server}

kilfoyle commented on docs/en/ingest-management/security/mutual-tls.asciidoc:
```diff
@@ -0,0 +1,200 @@
+|{agent} certificate to present to {fleet-server} during authentication
+
+|`-elastic-agent-cert-key`
+|{agent} certificate key to present to {fleet-server}
```
Changed to:

> -elastic-agent-cert-key
> {agent}'s private certificate key used to decrypt the certificate

kilfoyle commented on docs/en/ingest-management/security/mutual-tls.asciidoc:
```diff
@@ -0,0 +1,200 @@
+|{fleet-server} certificate to present to {agents} during authentication
+
+|`-fleet-server-cert-key`
+|{fleet-server} certificate key to present to {agents}  
```
Changed to:

> -fleet-server-cert-key
> {fleet-server}'s private certificate key used to decrypt the certificate

kilfoyle commented:
Thanks for the excellent review @AndersonQ! 🙏 
I've addressed all of your comments, most of them exactly as suggested.

AndersonQ commented on docs/en/ingest-management/security/mutual-tls.asciidoc:
```diff
@@ -0,0 +1,214 @@
+
+[cols="1,1"]
+|===
+|`ssl.certificate-authorities`
```
```suggestion
|`--certificate-authorities`
```

kilfoyle replied:
Fixed all instances.

AndersonQ commented on docs/en/ingest-management/security/mutual-tls.asciidoc:
```diff
@@ -0,0 +1,214 @@
+|`ssl.certificate-authorities`
+|List of CA certificates that are trusted when {fleet-server} connects to {agent} 
+
+|`-elastic-agent-cert`
```
```suggestion
|`--elastic-agent-cert`
```

kilfoyle replied:
Fixed all instances.

AndersonQ commented on docs/en/ingest-management/security/mutual-tls.asciidoc:
```diff
@@ -0,0 +1,214 @@
+|`-elastic-agent-cert`
+|{agent} certificate to present to {fleet-server} during authentication
+
+|`-elastic-agent-cert-key`
```
```suggestion
|`--elastic-agent-cert-key`
```

kilfoyle replied:
Fixed all instances.

AndersonQ reviewed (changes requested):
@kilfoyle my bad here, I should have explained better about the configuration name/key/flag. 

Actually I've just realised the design doc was wrong as well, it's fixed now.

When it's a cli parameter or flag, it's in the form `--`+`name-of-the-flag` (`--name-of-the-flag`).

For the policy, it's a YAML path like `some.yaml.path.to.my_config_name`. 

Below are the correct forms

### tl;dt

 - in the policy (the one for the advanced YAML box):
```
ssl.certificate_authorities:
  - /path/to/ca
ssl.certificate: /path/to/cert
ssl.key: /path/to/cert_key

# OR

  ssl:
    certificate_authorities:
      - /path/to/ca
    certificate: /path/to/cert
    key: /path/to/cert_key
```

- for the cli (install/enroll cmd)
```
--certificate-authorities
--elastic-agent-cert
--elastic-agent-cert-key

--certificate-authorities # this is the same cli flag as the one above
--fleet-server-cert
--fleet-server-cert-key

--fleet-server-es-ca
--fleet-server-es-cert
--fleet-server-es-cert-key
```

I suggested a fix for the 1st set of cli flags and there are 2 other minor comments

kilfoyle commented on docs/en/ingest-management/security/mutual-tls.asciidoc:
```diff
@@ -0,0 +1,214 @@
+
+. mTLS is enabled.
+. The initial mTLS connection between {agent} and {fleet-server} is configured when {agent} is enrolled, using the parameters passed through the `elastic-agent install` or `elastic-agent enroll` command.
+. Once enrollment is completed, {agent} downloads the initial agent policy from {fleet-server}. That policy contains the mTLS configuration settings that will take precedence over those used during enrollment: This includes both the mTLS settings used for connectivity between {agent} and {fleet-server} (and the {fleet} application in {kib}, for {fleet}-managed (agent)), and the settings used betw ...
```
Definitely worth mentioning! I've re-organized it a little bit:

![Screenshot 2024-06-20 at 11 50 31 AM](https://github.com/elastic/ingest-docs/assets/41695641/a14628ea-5840-40e1-932b-8193d6cdb3b7)

kilfoyle commented on docs/en/ingest-management/security/mutual-tls.asciidoc:
```diff
@@ -0,0 +1,214 @@
+
+In some scenarios you may want to deploy {fleet-server} on your own premises. In this case, you're able to provide your own certificates and certificate authority to enable mTLS between {fleet-server} and {agent}.
+
+However, as with the <<mutual-tls-cloud,{fleet-server} on {ecloud}>> use case, the data plane connections are has its TLS connection terminated at the {ecloud} boundary. {ecloud} is not a multi-tenanted service and therefore can't provide per-user certificates.  
```
I think "connections have their TLS connection" is also a bit confusing, so I've rewritten to:

![Screenshot 2024-06-20 at 11 53 39 AM](https://github.com/elastic/ingest-docs/assets/41695641/717101e1-7930-469e-a2be-744c1716dd31)

kilfoyle commented:
@AndersonQ this is great. I've fixed up all of the settings as you suggested.

For the policy settings that go in the "advanced YAML" box, I updated the screen capture so the example uses the correct format, and I also added the following in that section of the page, so that people will be sure to know what format to use:

![Screenshot 2024-06-20 at 11 57 30 AM](https://github.com/elastic/ingest-docs/assets/41695641/44e0493f-5df7-4f1e-8d58-771027e8c227)

kilfoyle commented:
@AndersonQ  When you have some time can you please give this one a final check?

nimarezainia commented:
@AndersonQ in context of this [issue](https://github.com/elastic/elastic-agent/issues/5037#issuecomment-2205066012) - could you please provide the updated cli flag description that you envisage so that the docs can reference the ideal description text and whoever takes over the issue can update the helper text according to what we place in the docs.

cc: @pierrehilbert

pierrehilbert commented:
@AndersonQ is on SDH rotation this week, unless the week is really noisy from that perspective, he should have time to do the final check soon.

AndersonQ commented:
sorry for the delay, the SDHs were keeping me quite busy, it's approved now.

pierrehilbert commented:
Thanks @AndersonQ 
@kilfoyle we are now good to go.

AndersonQ commented:
> @AndersonQ in context of this [issue](https://github.com/elastic/elastic-agent/issues/5037#issuecomment-2205066012) - could you please provide the updated cli flag description that you envisage so that the docs can reference the ideal description text and whoever takes over the issue can update the helper text according to what we place in the docs.
> 
> cc: @pierrehilbert

I updated https://github.com/elastic/elastic-agent/issues/5037 with a proposed changes and opened https://github.com/elastic/elastic-agent/pull/5048

kilfoyle commented:
Brilliant! Thanks a lot @AndersonQ!
